{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rosalind_9438 Rosalind_9063\n",
      "Rosalind_9438 Rosalind_6124\n",
      "Rosalind_9438 Rosalind_1088\n",
      "Rosalind_9438 Rosalind_9348\n",
      "Rosalind_5312 Rosalind_7612\n",
      "Rosalind_9512 Rosalind_6304\n",
      "Rosalind_2721 Rosalind_2830\n",
      "Rosalind_2721 Rosalind_9340\n",
      "Rosalind_6780 Rosalind_6492\n",
      "Rosalind_0727 Rosalind_2721\n",
      "Rosalind_3315 Rosalind_6166\n",
      "Rosalind_8534 Rosalind_6492\n",
      "Rosalind_1288 Rosalind_1162\n",
      "Rosalind_1288 Rosalind_7020\n",
      "Rosalind_1288 Rosalind_4076\n",
      "Rosalind_3065 Rosalind_9063\n",
      "Rosalind_3065 Rosalind_6124\n",
      "Rosalind_3065 Rosalind_1088\n",
      "Rosalind_3065 Rosalind_9348\n",
      "Rosalind_7582 Rosalind_6872\n",
      "Rosalind_7612 Rosalind_5312\n",
      "Rosalind_7612 Rosalind_1510\n",
      "Rosalind_7612 Rosalind_5585\n",
      "Rosalind_7612 Rosalind_1092\n",
      "Rosalind_8378 Rosalind_3254\n",
      "Rosalind_8378 Rosalind_8471\n",
      "Rosalind_6016 Rosalind_6150\n",
      "Rosalind_5799 Rosalind_1272\n",
      "Rosalind_8618 Rosalind_6166\n",
      "Rosalind_6480 Rosalind_8534\n",
      "Rosalind_6480 Rosalind_2427\n",
      "Rosalind_9063 Rosalind_5161\n",
      "Rosalind_9063 Rosalind_7796\n",
      "Rosalind_3038 Rosalind_9736\n",
      "Rosalind_3038 Rosalind_1989\n",
      "Rosalind_3038 Rosalind_0809\n",
      "Rosalind_3254 Rosalind_2721\n",
      "Rosalind_1581 Rosalind_0727\n",
      "Rosalind_1581 Rosalind_4819\n",
      "Rosalind_6124 Rosalind_8933\n",
      "Rosalind_6124 Rosalind_9414\n",
      "Rosalind_6124 Rosalind_5541\n",
      "Rosalind_8127 Rosalind_3038\n",
      "Rosalind_8127 Rosalind_9040\n",
      "Rosalind_1162 Rosalind_6480\n",
      "Rosalind_1162 Rosalind_8127\n",
      "Rosalind_1162 Rosalind_8082\n",
      "Rosalind_1162 Rosalind_0479\n",
      "Rosalind_1510 Rosalind_6780\n",
      "Rosalind_1510 Rosalind_1288\n",
      "Rosalind_1510 Rosalind_3065\n",
      "Rosalind_1510 Rosalind_8618\n",
      "Rosalind_1510 Rosalind_3034\n",
      "Rosalind_1510 Rosalind_0630\n",
      "Rosalind_1510 Rosalind_9380\n",
      "Rosalind_8671 Rosalind_5043\n",
      "Rosalind_8671 Rosalind_6517\n",
      "Rosalind_9736 Rosalind_6166\n",
      "Rosalind_2212 Rosalind_4900\n",
      "Rosalind_2212 Rosalind_4875\n",
      "Rosalind_2212 Rosalind_0539\n",
      "Rosalind_0070 Rosalind_1554\n",
      "Rosalind_4819 Rosalind_0103\n",
      "Rosalind_9722 Rosalind_7612\n",
      "Rosalind_5161 Rosalind_5665\n",
      "Rosalind_5161 Rosalind_7781\n",
      "Rosalind_5161 Rosalind_3791\n",
      "Rosalind_0693 Rosalind_2830\n",
      "Rosalind_0693 Rosalind_9340\n",
      "Rosalind_9554 Rosalind_9438\n",
      "Rosalind_9554 Rosalind_1843\n",
      "Rosalind_4792 Rosalind_0103\n",
      "Rosalind_1554 Rosalind_3315\n",
      "Rosalind_1554 Rosalind_2277\n",
      "Rosalind_1279 Rosalind_9438\n",
      "Rosalind_1279 Rosalind_1843\n",
      "Rosalind_4032 Rosalind_5161\n",
      "Rosalind_4032 Rosalind_7796\n",
      "Rosalind_1989 Rosalind_7582\n",
      "Rosalind_2427 Rosalind_8328\n",
      "Rosalind_3034 Rosalind_9722\n",
      "Rosalind_0630 Rosalind_2830\n",
      "Rosalind_0630 Rosalind_9340\n",
      "Rosalind_2162 Rosalind_3070\n",
      "Rosalind_9578 Rosalind_7582\n",
      "Rosalind_8933 Rosalind_7582\n",
      "Rosalind_9380 Rosalind_9438\n",
      "Rosalind_9380 Rosalind_1843\n",
      "Rosalind_6166 Rosalind_2830\n",
      "Rosalind_6166 Rosalind_9340\n",
      "Rosalind_5043 Rosalind_6780\n",
      "Rosalind_5043 Rosalind_1288\n",
      "Rosalind_5043 Rosalind_3065\n",
      "Rosalind_5043 Rosalind_8618\n",
      "Rosalind_5043 Rosalind_3034\n",
      "Rosalind_5043 Rosalind_0630\n",
      "Rosalind_5043 Rosalind_9380\n",
      "Rosalind_9340 Rosalind_3254\n",
      "Rosalind_9340 Rosalind_8471\n",
      "Rosalind_1088 Rosalind_9512\n",
      "Rosalind_6492 Rosalind_8378\n",
      "Rosalind_5665 Rosalind_5043\n",
      "Rosalind_5665 Rosalind_6517\n",
      "Rosalind_0026 Rosalind_6480\n",
      "Rosalind_0026 Rosalind_8127\n",
      "Rosalind_0026 Rosalind_8082\n",
      "Rosalind_0026 Rosalind_0479\n",
      "Rosalind_9082 Rosalind_3242\n",
      "Rosalind_9082 Rosalind_1279\n",
      "Rosalind_9082 Rosalind_9578\n",
      "Rosalind_0539 Rosalind_7460\n",
      "Rosalind_0809 Rosalind_8328\n",
      "Rosalind_6872 Rosalind_0026\n",
      "Rosalind_6872 Rosalind_3388\n",
      "Rosalind_7460 Rosalind_1272\n",
      "Rosalind_3249 Rosalind_7582\n",
      "Rosalind_6150 Rosalind_8378\n",
      "Rosalind_9348 Rosalind_3242\n",
      "Rosalind_9348 Rosalind_1279\n",
      "Rosalind_9348 Rosalind_9578\n",
      "Rosalind_7796 Rosalind_5312\n",
      "Rosalind_7796 Rosalind_1510\n",
      "Rosalind_7796 Rosalind_5585\n",
      "Rosalind_7796 Rosalind_1092\n",
      "Rosalind_7781 Rosalind_9723\n",
      "Rosalind_7781 Rosalind_8671\n",
      "Rosalind_7781 Rosalind_4792\n",
      "Rosalind_9414 Rosalind_0103\n",
      "Rosalind_4486 Rosalind_3254\n",
      "Rosalind_4486 Rosalind_8471\n",
      "Rosalind_2347 Rosalind_3070\n",
      "Rosalind_3388 Rosalind_8302\n",
      "Rosalind_3388 Rosalind_1581\n",
      "Rosalind_1833 Rosalind_5312\n",
      "Rosalind_1833 Rosalind_1510\n",
      "Rosalind_1833 Rosalind_5585\n",
      "Rosalind_1833 Rosalind_1092\n",
      "Rosalind_3791 Rosalind_9736\n",
      "Rosalind_3791 Rosalind_1989\n",
      "Rosalind_3791 Rosalind_0809\n",
      "Rosalind_6304 Rosalind_1162\n",
      "Rosalind_6304 Rosalind_7020\n",
      "Rosalind_6304 Rosalind_4076\n",
      "Rosalind_8082 Rosalind_2212\n",
      "Rosalind_7955 Rosalind_7582\n",
      "Rosalind_6517 Rosalind_5043\n",
      "Rosalind_4076 Rosalind_1554\n",
      "Rosalind_9040 Rosalind_5161\n",
      "Rosalind_9040 Rosalind_7796\n",
      "Rosalind_0479 Rosalind_3038\n",
      "Rosalind_0479 Rosalind_9040\n",
      "Rosalind_1843 Rosalind_8378\n"
     ]
    }
   ],
   "source": [
    "#grph\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "sequences = []\n",
    "ids = []\n",
    "counter_s = 0\n",
    "counter_t = 0\n",
    "\n",
    "with open('/Users/nocide51/Projects_Coding/psc2k22/asignment5/sequences/rosalind_grph.txt', 'r') as handle:\n",
    "    for x in SeqIO.parse(handle, \"fasta\"):\n",
    "        sequences.append(x)\n",
    "        ids.append(x.id)\n",
    "    for seq in sequences:\n",
    "        counter_t = 0\n",
    "        s = list(seq[-3:])\n",
    "        s_id = ids[counter_s]\n",
    "        #print(s)\n",
    "        for seq2 in sequences:\n",
    "            t = list(seq2[:3])\n",
    "            t_id = ids[counter_t]\n",
    "            #print(t)\n",
    "            if s == t:\n",
    "                if s_id == t_id:\n",
    "                    continue\n",
    "                else:\n",
    "                    print(s_id + ' ' + t_id)\n",
    "                \n",
    "            counter_t += 1\n",
    "        counter_s += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Q05557', 'P81448', 'P04141', 'P07306', 'B5FNU0', 'A9QYR8', 'P55067', 'P01044', 'B4S2L7', 'A3N0C7', 'P42098', 'P04278', 'Q50228']\n",
      "Q05557\n",
      "\n",
      "P81448\n",
      "86\n",
      "P04141\n",
      "44 54\n",
      "P07306\n",
      "79 147\n",
      "B5FNU0\n",
      "\n",
      "A9QYR8\n",
      "183 210\n",
      "P55067\n",
      "121 339 737 967 1164\n",
      "P01044\n",
      "47 87 168 169 197 204\n",
      "B4S2L7\n",
      "33\n",
      "A3N0C7\n",
      "59\n",
      "P42098\n",
      "124 146 179 271\n",
      "P04278\n",
      "380 396\n",
      "Q50228\n",
      "55 228\n"
     ]
    }
   ],
   "source": [
    "#mprt\n",
    "\n",
    "\n",
    "import requests as r\n",
    "from Bio import SeqIO\n",
    "from io import StringIO\n",
    "import re\n",
    "\n",
    "cID=[]\n",
    "#motif = 'N[^P][ST][^P]'\n",
    "motif = re.compile(r'(?=(N[^P][ST][^P]))') \n",
    "file = open('/Users/nocide51/Projects_Coding/psc2k22/asignment5/sequences/rosalind_mprt.txt', 'r')\n",
    "for line in file:\n",
    "    if '_' in line:\n",
    "        line = line.split('_')\n",
    "        cID.append(line[0])\n",
    "    else:\n",
    "        cID.append(line.strip())\n",
    "print(cID)\n",
    "\n",
    "\n",
    "baseUrl=\"http://www.uniprot.org/uniprot/\"\n",
    "for id in cID:\n",
    "    currentUrl=baseUrl+id+\".fasta\"\n",
    "    response = r.post(currentUrl)\n",
    "    cData=''.join(response.text)\n",
    "    Seq=StringIO(cData)\n",
    "    pSeq=list(SeqIO.parse(Seq,'fasta'))\n",
    "    #print(pSeq)\n",
    "    my_seq = str(pSeq[0].seq)\n",
    "    #print(my_seq)\n",
    "    print(id)\n",
    "    positions = re.findall(motif, my_seq)\n",
    "    #if positions == []:\n",
    "        #continue\n",
    "    #else:\n",
    "        #print(positions)\n",
    "    result = []\n",
    "    for glyco in positions:  \n",
    "        if glyco in my_seq:\n",
    "            result.append(my_seq.index(glyco)+1)\n",
    "    print(*result)\n",
    "\n",
    "        \n",
    "        \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MMQTDRLHYSTELTSK\n",
      "MQTDRLHYSTELTSK\n",
      "MSVDPLHAAPTAFG\n",
      "MLSSMYEGVPTDPSTSLRMLYPSR\n",
      "MYEGVPTDPSTSLRMLYPSR\n",
      "MRGCLQIHQLVCGCCIPAVNLVHLTHSYWRRVIDKARSF\n",
      "MLYPSR\n",
      "MHVRGRRRLAKRRRPRTCGERDTTRVSRQLSKNTEVCPCVDRERNSGSCIVGLWDNQEDWHLILP\n",
      "MSAAGGA\n",
      "MRGKRYN\n",
      "MIHDYPLSYRVGGRSIR\n",
      "MLLRAKRNDPSNTDPFPTLSVGSLPS\n",
      "MPHWHHPPP\n",
      "MSGTGRCYSDRSAWPEEA\n",
      "MDRPPTLYDKG\n",
      "MIKGNRGSFRNGLGDPPDQALYVRCPHGRMRCQSS\n",
      "MSVALTVE\n",
      "MRCQSS\n",
      "MQEPEFRSRSTQGQTSVFFES\n",
      "MHQMELLALFVEGR\n",
      "MELLALFVEGR\n",
      "MDL\n",
      "MLESM\n",
      "M\n",
      "MVHVRIS\n"
     ]
    }
   ],
   "source": [
    "#orf\n",
    "\n",
    "import Bio\n",
    "pattern = re.compile(r'(?=(ATG(?:...)*?)(?=TAG|TGA|TAA))')\n",
    "\n",
    "def find_orf(frame):\n",
    "    for x in re.findall(pattern, str(frame)):\n",
    "        dna = Seq(x)\n",
    "        prot = dna.translate()\n",
    "        if prot not in sequences:\n",
    "            sequences.append(prot)\n",
    "\n",
    "\n",
    "\n",
    "with open('/Users/nocide51/Projects_Coding/psc2k22/asignment5/sequences/rosalind_orf.txt', 'r') as handle:\n",
    "    sequence = []\n",
    "    sequences = []\n",
    "    for x in SeqIO.parse(handle, \"fasta\"):\n",
    "        sequence.append(x)\n",
    "    my_sequence = sequence[0].seq\n",
    "    my_reverse_aa = my_sequence.reverse_complement()\n",
    "    my_aa = my_sequence\n",
    "    #frames = []\n",
    "    #frames.append(str(my_aa.seq))\n",
    "    #frames.append(str(my_reverse_aa.seq[1:]))\n",
    "    #frames.append(str(my_reverse_aa.seq[2:]))\n",
    "    #frames.append(str(my_aa.seq[1:]))\n",
    "    #frames.append(str(my_aa.seq[2:]))\n",
    "    #print(frames)\n",
    "    #for x in range(len(frames)):\n",
    "        #print(find_orf(frames[x]))\n",
    "    find_orf(my_reverse_aa)\n",
    "    find_orf(my_aa)\n",
    "    for a, b in enumerate(sequences):\n",
    "        #print(a)\n",
    "        print(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGATAGCAGCATACAGTGAAATCATTTCTTCCTCTGGATTCGTGAATAGGCGGGGGGCTGATTGCCCTCTGGTCAGCTTGGCCCCATTGTCCGAAGGATTGTCGAACCCCACACTCACCACAGATCGCAACAGTGCATCCAAGTAAGTAACATATCTCGTGTGCTTCGGTCGGGCTGGGGGTTCTTCATAACCCCACCTCTGCCGACCGCCACCCGGCGAAAGCAACCAGCAGGGTTCGTCGAATGCTAGTTTACTGTGTGATTCCATATTATACGACGGGAACAGGACGATGGACCGCATTAGCGATCAGGATCGAATAGCTCCATCGGAGAGCCAACCATTGGGTTCTCATCAATCGGTAGACCCCTGGTATCGACTGGGTGGATGCGCGACTGTATACTCACATAGTTCACCAAGGATTTTCTTCACCAAAATCGTACATCGACATTTCCTGTACACGTAGTGCCCAGTAACAGTGCATAAAAATGAGTCGTCAATTTATTTCGGTACAGGCAGAACGTTGTGGGTGCTCTAGCCTACAAGCAGCCAGGACCGAACCGACGCGGTACGGATTAATCGTTTCTCTCGCCATCTTTCTAGCAACGGCTTAGCTACCTTGGCTTTATCTCCTAACATCCTTAATAGAACAGGGCAATACATTTTACTGATGTACATTGCACCTTGCCGTATTTTAACCTTGGTGAGGTCTGACGCGAGTCCCATAACAGGACCACTGGTGTGAGGTACGGGGCGAGTCATCTCGGAGTGATATTGCACCCTTCCGCTGTTCCTTACGGGGAATATCATGGACTCTCGGTTTGCTCAGAGCGGACCGGCTTGCCGGTGGCTCTCCGCACAAATGCGAAAGCCCTCTGCGCGGATGAACAAATACATTTCGAGAAGGGTCCCGATTCGACTTGAGGTGGGTTTACGAACTGCACGAGAGTCTAGTGAACGCCTCTACCCATTTGA\n",
      "[Seq('GCTCTCCGCACAAAT'), Seq('TGAGGTGGGTTTACGAACTGCACGAGAGTCTAGTGA'), Seq('ACAGTGCATCCAAGTAAGTAA'), Seq('AACAGTGCATAAAAATGAGT'), Seq('GGATGCGCGACTGTATACTCACATAGTTCACCAAGGATTTTCTTCACCAA'), Seq('TCTCGGAGTGATATTGCACCCTTCCGCTGTTCCTTAC'), Seq('ATTTTAACCTTGGTGAGGTCTGACGCGAGTCCCATAACAGGAC'), Seq('AGGACCGAACCG'), Seq('GCCATCTTTCTAGCAACGGCTTAG'), Seq('CGGGGGGCTGATTGCCCTCTGGTCAGCTTGGCCCCATTGTCCG'), Seq('TTAATAGAACAGGGCAATACATTTTACTG'), Seq('GATCAGGATCGAATAGCTCCATCGGAGAGCCAACCATTGGGTTCTCAT'), Seq('CACCCGGCGAAAGCAACCAGCAGGGTTCGTCGAATGCTAGT')]\n",
      "ATGATAGCAGCATACAGTGAAATCATTTCTTCCTCTGGATTCGTGAATAGGAAGGATTGTCGAACCCCACACTCACCACAGATCGCACATATCTCGTGTGCTTCGGTCGGGCTGGGGGTTCTTCATAACCCCACCTCTGCCGACCGCTTACTGTGTGATTCCATATTATACGACGGGAACAGGACGATGGACCGCATTAGCCAATCGGTAGACCCCTGGTATCGACTGGGTAATCGTACATCGACATTTCCTGTACACGTAGTGCCCAGTCGTCAATTTATTTCGGTACAGGCAGAACGTTGTGGGTGCTCTAGCCTACAAGCAGCCACGCGGTACGGATTAATCGTTTCTCTCCTACCTTGGCTTTATCTCCTAACATCCATGTACATTGCACCTTGCCGTCACTGGTGTGAGGTACGGGGCGAGTCAGGGGAATATCATGGACTCTCGGTTTGCTCAGAGCGGACCGGCTTGCCGGTGGCGAAAGCCCTCTGCGCGGATGAACAAATACATTTCGAGAAGGGTCCCGATTCGACTACGCCTCTACCCATTTGA\n",
      "MIAAYSEIISSSGFVNRKDCRTPHSPQIAHISCASVGLGVLHNPTSADRLLCDSILYDGNRTMDRISQSVDPWYRLGNRTSTFPVHVVPSRQFISVQAERCGCSSLQAATRYGLIVSLLPWLYLLTSMYIAPCRHWCEVRGESGEYHGLSVCSERTGLPVAKALCADEQIHFEKGPDSTTPLPI*\n"
     ]
    }
   ],
   "source": [
    "#splc\n",
    "import Bio\n",
    "\n",
    "def spliceosome(dna, intron):\n",
    "    dna = str(dna)\n",
    "    intron = str(intron)\n",
    "    if intron in dna:\n",
    "        start = dna.index(intron)\n",
    "        length = len(intron)\n",
    "        end = start+length\n",
    "        first_part = dna[:start]\n",
    "        intron_part = dna[start:end]\n",
    "        second_part = dna[end:]\n",
    "        without_intron = first_part + second_part\n",
    "        #print(without_intron)\n",
    "        return without_intron\n",
    "\n",
    "\n",
    "\n",
    "sequences = []\n",
    "with open('/Users/nocide51/Projects_Coding/psc2k22/asignment5/sequences/rosalind_splc.txt', 'r') as handle:\n",
    "    for x in SeqIO.parse(handle, 'fasta'):\n",
    "        sequences.append(x.seq)\n",
    "    dna = sequences[0]\n",
    "    print(dna)\n",
    "    #spliceosome(dna, sequences[1])\n",
    "    print(sequences[1:])\n",
    "    new_dna = []\n",
    "    for x in range(1, len(sequences)):\n",
    "        if x == 1:\n",
    "            new_dna.append(spliceosome(dna, sequences[x]))\n",
    "        else:\n",
    "           new_dna.append(spliceosome(new_dna[-1], sequences[x]))\n",
    "    print(new_dna[-1])\n",
    "    prot_seq = Seq(new_dna[-1]).translate()\n",
    "    print(prot_seq)\n",
    "        "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinformatics",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "fee94040681bba65a2a9b072a34f3035c29fcc436a8139ce1c49f0a1ccfba704"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
